test/metabase/models/dashboard_test.clj

```diff
@@ -634,6 +634,62 @@
      (testing "there are no \"Tab 3\""
        (is (false? (t2/exists? :model/DashboardTab :dashboard_id dashboard-id :name "Tab 3")))))))
 
+(deftest revert-dashboard-skip-archived-or-deleted-card-test
+  (t2.with-temp/with-temp [:model/Dashboard {dashboard-id :id}          {:name "A dashboard"}
+                           :model/Card      {will-be-archived-card :id} {}
+                           :model/Card      {will-be-deleted-card :id}  {}
+                           :model/Card      {unchanged-card       :id}  {}]
+    ;; 0. create a dashboard
+    (create-dashboard-revision! dashboard-id true)
+    ;; 1. add 3 cards and 1 text card
+    (t2/insert-returning-pks! :model/DashboardCard
+                              [{:dashboard_id     dashboard-id
+                                :dashboard_tab_id nil
+                                :card_id          will-be-archived-card
+                                :row              0
+                                :col              0
+                                :size_x           4
+                                :size_y           4}
+                               {:dashboard_id     dashboard-id
+                                :dashboard_tab_id nil
+                                :card_id          will-be-deleted-card
+                                :row              4
+                                :col              4
+                                :size_x           4
+                                :size_y           4}
+                               {:dashboard_id     dashboard-id
+                                :dashboard_tab_id nil
+                                :card_id          unchanged-card
+                                :row              0
+                                :col              0
+                                :size_x           4
+                                :size_y           4}
+                               {:dashboard_id           dashboard-id
+                                :dashboard_tab_id       nil
+                                :row                    4
+                                :col                    4
+                                :size_x                 4
+                                :size_y                 4
+                                :visualization_settings {:text "Metabase"}}])
+    (create-dashboard-revision! dashboard-id false)
+
+    ;; 2. delete all the dashcards
+    (t2/delete! :model/DashboardCard :dashboard_id dashboard-id)
+    (create-dashboard-revision! dashboard-id false)
+
+    ;; cards can be deleted if its database is deleted
```

Comment:
Cards can also be deleted if the question is archived and then the question is deleted from the archive. This comment
confused me because its not relevant to the test and it can get out of date. I think you can remove it.